client/concordclient/include/client/concordclient/remote_update_queue.hpp

```diff
@@ -0,0 +1,129 @@
+// Concord
+//
+// Copyright (c) 2020-2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <condition_variable>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <exception>
+
+#include "assertUtils.hpp"
+
+#include "client/concordclient/remote_update_data.hpp"
+
+namespace concord::client::concordclient {
+
+// An implementation of RemoteUpdateQueue should guarantee that ReleaseConsumers, Clear, Push, Pop, and TryPop are all
+// mutually thread safe.
+template <typename T>
+class RemoteUpdateQueue {
```

Comment:
Thanks, @robem, as UpdateQueue is a data structure required for Inter Thread Communication, this name is changed. A better name for this queue should be LockLess or ThreadSafeQueue and not UpdateQueue. 
RemoteUpdateQueue is used since it signifies that a remotely running thread can Update this Queue and synchronize with other threads.
Please let me know what you think, can I name it as ThreadSafeQueue?

Reply:
This is a good point, what is a good name for this class, that you suggest. In my opinion, UpdateQueue is also not nice, but I will go with your suggestion. In the interface comments, we can see that it's saying that the Implementation should guarantee thread safety as its purpose is to use it in concurrently running threads. So I thought ThreadSafeQueue might be better so that the implementor can always make a thread-safe implementation. But again it would be good if you can provide a better name for this with a second thought.

Reply:
thanks, @robem , I am keeping the name of the interface as IQueue.